Feat/#83

jyj1289 opened the pull request:
## 🎫 관련 이슈
[//]: # (다음 키워드를 사용하면 해당 PR을 머지할 때 자동으로 이슈를 닫을 수 있습니다.)
[//]: # (keyword: close|closes|closed|resolve|resolves|resolved|fix|fixes|fixed)
[//]: # (예시: close #1)

close #83 

<br>

## 📄 개요
[//]: # (작업 내용을 간단히 요약해서 적습니다.)
[//]: # (예시: 유저 회원가입 기능을 만들었습니다.)

> 유저 비밀번호 변경 기능을 만들었습니다.

<br>

## 🔨 작업 내용
[//]: # (작업 내용을 자세하게 적습니다.)
[//]: # (붙임표 "-" 을 사용해서 목록을 만듭니다.)
[//]: # (예시: 유저 회원가입 API를 만들었습니다.)

- 유저 비밀번호 변경 API를 만들었습니다.
- 유저 클래스에 비밀번호 변경 메서드를 추가했습니다.
- 유저 생성 유즈케이스의 내용을 변경하였습니다.


<br>

## 🏁 확인 사항
[//]: # (PR을 보내기 전 다음 사항을 확인해주세요.)
[//]: # (해당 사항을 모두 이행해야 머지할 수 있습니다.)
[//]: # (- [x] 를 사용해서 완료로 표시할 수 있습니다.)

- [x] 테스트를 완료했나요?
- [x] API 문서를 작성했나요?
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 로그, 주석, import 등을 삭제했나요?

<br>

## 🙋🏻 덧붙일 말
[//]: # (다음 사항이 있다면 적어주세요.)
[//]: # (PR에 대한 추가 설명)
[//]: # (중점적으로 리뷰받고 싶은 부분)
[//]: # (기타 등등)

jyj1289 committed:
feat : MessageController
- 메시지를 그룹별로 보낼 수 있는 메세지 컨트롤러를 만들었습니다.

jyj1289 committed:
feat(#75):send message by group service
- 메시지를 그룹별로 보낼 수 있는 메세시를 보낼 수 있는 서비스를 만들었습니다.

jyj1289 committed:
docs:add cors url
- cors에 마루 버셀 url을 추가했습니다.

jyj1289 committed:
feat(#75):MessageControllerTest
- MessageControllerTest를 만들었어요.

jyj1289 committed:
docs(#75):MessageController asciidocs
- MessageController에 대한 아스키 독스를 만들었어요

jyj1289 committed:
fix(#75) : remove verify()
- 횟수 검증을 할 필요가 없어서 verify() 호출 코드를 제거했어요

jyj1289 committed:
refactor(#75) : 주석 제거
- 불필요한 주석을 제거했어요

jyj1289 committed:
refactor(#75) : 중복 코드 제거
- 불필요한 중복 코드를 제거했어요

jyj1289 committed:
refactor(#75) : 자동 정렬
- 코드를 정렬했어요

jyj1289 committed:
feat(#75) : SendMessageUseCase
- SendMessageUseCase를 만들었어요

jyj1289 committed:
refact(#75) : MessageUseCase 사용
- execute에서 phoneNumberList를 생성하던 것을 UseCase로 이전했어요.
- 기존 for문을 stream으로 변환했어요.

jyj1289 committed:
refact(#75) : MessageUseCase 사용
- SendMessageService를 사용하던 코드들을 SendMessageUseCase를 사용하도록 변환했어요.

jyj1289 committed:
refactor(#75) : stream 컨벤션 사용
- stream 컨벤션을 사용했어요.

jyj1289 committed:
refactor(#75) : sendMessage 메서드 이름 변경 및 엔드포인트 변경
- 원서 상태별로 메시지 보내기와 원서 타입별로 메시지 보내기 기능을 분리하기 위해 이름과 엔드포인트를 변경했어요.

jyj1289 committed:
test(#75) : 엔드포인트 변경 및 BDDMockito 메서드로 변경
- BDD관점으로 보기 위해서 BDDMockito메서드로 변경했어요.
- UseCaseTest에서 status별로 테스트하기 위해서 컨트롤러 테스트를 수정했어요.

jyj1289 committed:
test(#75) : status별 단위 테스트 작성
- SendMessageUseCaseTest를 만들었어요.

jyj1289 committed:
Merge pull request #80 from Bamdoliro/feat/#75

Feat/#75

jyj1289 committed:
feat(#81) : 불필요한 코드 개선
- SendMessageService에서 formRepository를 사용하지 않아서 제거했어요.

jyj1289 committed:
feat(#81) : 메서드 추가
- 마이스터전형과, 마이스터전형을 제외한 모든 전형을 조회하기 위해서 findMeisterTalentFirstRoundForm()과 findNotExistsMeisterTalentFirstRoundForm() 메서드를 추가했어요

jyj1289 committed:
feat(#81) : formRepository 메서드 구현
- 마이스터전형과, 마이스터전형을 제외한 모든 전형을 조회하기 위해서 findMeisterTalentFirstRoundForm()과 findNotExistsMeisterTalentFirstRoundForm() 메서드를 구현했어요.

jyj1289 committed:
feat(#81) : 전형별 메세지 메서드 추가
- 전형별로 메시지를 보내기 위해서 컨트롤러에 메서드를 추가했어요.

jyj1289 committed:
refactor(#81) : DTO 이름 변경
- SendMessageByStatusRequest로 이름을 변경했어요

jyj1289 committed:
feat(#81) : execute 메서드 추가 및 개선
- 전형별 메시지 보내기 메서드를 만들었어요.
- 비즈니스 로직을 private 메서드로 바꾸었어요.
- 겹치는 로직을 하나의 메서드로 개선했어요

jyj1289 committed:
docs : message.adoc 변경

jyj1289 committed:
Merge branch 'develop' of https://github.com/Bamdoliro/marubase into feat/#81

jyj1289 committed:
feat(#81) : SendMessageByTypeRequest 추가
- SendMessageByTypeRequest DTO를 추가했어요.

jyj1289 committed:
fix(#81) : SendMessageRequest 이름 변경
- SendMessageRequest를 SendMessageByStatusRequest로 변경했어요

jyj1289 committed:
refactor(#81) : import 제거
- 불필요한 import를 제거했어요.

jyj1289 committed:
fix(#81) : 로직 수정
- 마이스터 전형에서 일반전형으로 바뀐 원서는 getType()을 사용했을 때만 REGULAR로 나오고 실제 디비에는 MEISTER_TALENT로 들어가서 그룹별 메시지를 보낼 때 번거로움이 있어서 원서 전형 자체를 REGULAR로 바꾸기로 결정했어요.

jyj1289 committed:
fix(#81) : boolean 타입 변경
- 요청할 때 인식이 되지 않아서 boolean 타입에서 Boolean 타입으로 바꾸었어요

jyj1289 committed:
docs(#81) : message.adoc
- 요청할 때 인식이 되지 않아서 boolean 타입에서 Boolean 타입으로 바꾸었어요

jyj1289 committed:
feat(#81) : messageControllerTest 추가
- /message/type 테스트를 추가했어요.

jyj1289 committed:
fix(#81) : 테스트 코드 삭제
- 테스트를 위한 println코드를 삭제했어요

jyj1289 committed:
feat(#81) : 전형별 메시지 보내기 테스트 코드 작성
- 전형별 메시지 보내기 테스트 코드를 작성했어요.

jyj1289 committed:
Merge pull request #82 from Bamdoliro/feat/#81

Feat/#81

jyj1289 committed:
refactor(#83) : 인증번호 전송 문구 변경
- 회원가입 뿐만 아니라 비밀번호 변경을 할 때도 인증번호를 사용해야해서 사용자의 혼란을 막기 위해 문구를 변경했어요.

jyj1289 committed:
refactor(#83) : 불필요한 import 제거
- 사용하지 않는 import를 제거했어요

jyj1289 committed:
feat(#83) : 비밀번호 변경 api 테스트 추가
- 비밀번호 변경 api 테스트 코드를 작성했어요.

jyj1289 committed:
fix(#83) : 전화번호 인증 메서드 변경
- 비밀번호를 변경할때도 전화번호 인증을 사용해야하기 때문에 인증이 성공한 경우 다시 isVerified를 false로 변경하게 만들었어요.

jyj1289 committed:
feat(#83) : UpdatePasswordRequest 추가
- UpdatePasswordRequest를 만들었어요.

jyj1289 committed:
feat(#83) : UpdatePasswordUseCase 추가
- UpdatePasswordUseCase를 만들었어요.

jyj1289 committed:
feat(#83) : UpdatePasswordUseCaseTest 추가
- UpdatePasswordUseCaseTest를 작성했어요

jyj1289 committed:
docs(#83) : 비밀번호 변경 메뉴얼 추가
- 비밀번호 변경 api 관한 메뉴얼을 작성했어요.

jyj1289 committed:
feat(#83) : updatePassword 메서드 추가
- 비밀번호 변경을 위해 User 클래스에 updatePassword 메서드를 추가했어요

jyj1289 committed:
feat(#83) : 비밀번호 변경 api 추가
- 비밀번호 변경 api를 추가했어요.

jyj1289 committed:
feat(#83) : 비밀번호 변경 api 테스트 추가
- 비밀번호 변경 api를 테스트 코드를 작성했어요.

gimhanul commented on src/main/java/com/bamdoliro/maru/application/user/SignUpUserUseCase.java:
```diff
@@ -45,5 +44,10 @@ private void validate(SignUpUserRequest request) {
             throw new UserAlreadyExistsException();
         }
+
+        verificationRepository.updateVerification(
```
p2) false 상태로 만들어주는 것보다 삭제하는 게 낫지 않을까용?

gimhanul replied:
추가로 Verificaiton 도메인 보면 유효 시간 걸려있는 거 참고해주세요!

gimhanul commented on src/main/java/com/bamdoliro/maru/application/user/SendVerificationUseCase.java:
```diff
@@ -17,7 +17,7 @@ public class SendVerificationUseCase {
         Verification verification = new Verification(request.getPhoneNumber());
         String text = String.format(
-                "[부산소프트웨어마이스터고] 회원가입 인증번호는 [%s]입니다.",
+                "[부산소프트웨어마이스터고] 인증번호는 [%s]입니다.",
```
P1) 개인적인 생각이긴 한데, 요런 건 **인증번호를 받는 목적을 정확하게 명시**해줘야 한다고 생각해요
만약 고의로 다른 사람의 전화번호를 입력한다면 생각지 못한 일이 생길수도 있으니까?요

또한, "회원가입"을 위한 인증번호 저장소와 "비밀번호 변경"을 위한 인증번호 저장소를 같은 곳을 사용하면 안 돼요
트랜잭션 관리가 된다면 문제 없을 거 같긴 하지만, 기억상 redis는 `@Transactional` 트랙잭션이 안 걸리는 걸로 알고 있거든요

그래서 **redis 트랜잭션을 관리하거나, 인증번호 별 저장소를 분리하는 식**으로 관리가 되어야 할 거 같습니다

gimhanul commented on src/main/java/com/bamdoliro/maru/application/user/UpdatePasswordUseCase.java:
```diff
@@ -0,0 +1,48 @@
+            throw new VerifyingHasFailedException();
+        }
+
+        if (!userRepository.existsByPhoneNumber(request.getPhoneNumber())) {
```
P2) 굳이 이 메서드를 부를 필요가 없을 거 같아요
쿼리문이 오히려 두 번 나가서 손해일 거 같습니다
`userFacade.getUser()` 메서드가 findById 후에 유저가 없다면 orElseThrow로 아마 exception 던져주고 있을 거거든요

gimhanul replied:
facade 한 번 뜯어보고 결정해주세요! 아닐 수도 있어서 ㅎㅎ

gimhanul commented on src/main/java/com/bamdoliro/maru/application/user/UpdatePasswordUseCase.java:
```diff
@@ -0,0 +1,48 @@
+            throw new UserNotFoundException();
+        }
+
+        verificationRepository.updateVerification(
```
p2) false 상태로 만들어주는 것보다 삭제하는 게 낫지 않을까용?

gimhanul commented on src/main/java/com/bamdoliro/maru/presentation/user/UserController.java:
```diff
@@ -63,4 +60,12 @@ public void verify(
     }
+
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @PutMapping("/password-update")
```
P1) Http method중 put과 patch의 차이점을 알아보고 잘 구분해서 쓰면 더 적절하게 설계할 수 있을 거 같아요
그리고, password-update처럼 url에 동사를 넣지 않는 게 나을 거 같아요 어차피 put이나 patch가 update한다는 사실을 충분히 설명하고 있으니까용

gimhanul reviewed (changes requested):
열심히 테스트하는 모습 좋습니다!

jyj1289 committed:
fix(#83) : 메서드 이름 완성
- 메서드 이름을 다 쓰지 않고 커밋하여 수정하였습니다.

jyj1289 committed:
refactor(#83) : 메서드 이름 및 레포지토리 이름변경
- SignUp과 Password로 레포지토리를 나누었기 때문에 메서드 이름과 레포지토리 이름들을 다 변경했어요.

jyj1289 committed:
feat(#83) : SendVerificationRequest 필드 추가
- SignUp과 Password 레포지토리를 분리하여 이를 구분할 수 있는 VerificationType를 필드로 추가했어요.

jyj1289 committed:
feat(#83) : SendVerificationUseCase 코드 추가
- 회원가입과 비밀번호 변경 인증번호를 따로 두기 위해서 if문으로 구분했어요.

jyj1289 committed:
feat(#83) : SignUpVerification
- SignUpVerification 클래스를 추가했어요.

jyj1289 committed:
feat(#83) : SignUpVerificationRepository
- SignUpVerificationRepository를 추가했어요.

jyj1289 committed:
feat(#83) : UpdatePasswordVerificationRepository
- UpdatePasswordVerificationRepository를 추가했어요.

jyj1289 committed:
feat(#83) : VerificationRedisRepository 메서드 수정
- VerificationRedisRepository에 updateVerification을 SignUp과 Password로 구분했어요

jyj1289 committed:
refactor(#83) : VerificationRepository 삭제
- VerificationRepository을 삭제했어요

jyj1289 committed:
feat(#83) : updateVerification 메서드 변경
- signup과 password로 구분할 수 있게 메서드를 2개로 나누었어요.

jyj1289 committed:
feat(#83) : VerificationType
- SignUp인지 Password인지 확인하기 위한 Enum을 만들었어요.

jyj1289 committed:
feat(#83) : VerificationType 추가
- 회원가입과 비밀번호 변경을 구분하기 위해서 request에 type을 추가했어요.

jyj1289 committed:
feat(#83) : 회원가입과 비밀번호 변경을 구분하도록 변경
- 회원가입과 비밀번호 변경을 구분해서 인증번호를 저장하도록 바꾸었어요.

jyj1289 committed:
feat(#83) : EnumController에 등록
- EnumController에 새로 마든 인증 타입을 추가했어요.

jyj1289 committed:
feat(#83) : EnumDocs에 등록
- EnumDocs에 새로 마든 인증 타입을 추가했어요.

jyj1289 committed:
refactor(#83) : 이름 다 변경 및 request 수정
- 회원가입과 비밀번호 변경인증코드를 구분하여 다른 메서드 내부에서도 구분하도록 변경하고, request에 전달인자 내용도 바꾸었어요.

jyj1289 committed:
refactor(#83) : 이름 변경
- 메서드 이름을 변경하고 사용하는 메서드 이름도 변경했어요.

jyj1289 committed:
fix(#83) : 메서드 변경 밑 api 엔드포인트 변경
- 메서드를 patch로 변경하고 엔드포인트를 password로 변경했어요.

jyj1289 committed:
fix(#83) : EnumController에 추가
- VerificationType을 EnumController 및 SharedController에 추가했어요.
- 추가하기 위해서 VerificationType에 EnumProperty를 구현했어요.

jyj1289 committed:
feat(#83) : 회원가입 인증 번호 및 비밀번호 변경 인증 테스트
- 회원가입과 비밀번호 변경 인증번호를 구분하여 테스트 하는 메서드를 추가했어요.

jyj1289 committed:
feat(#83) : 비밀번호 변경 인증 테스트 추가
- 비밀번호 변경 인증번호를 요청, 완료하는 테스트를 추가했어요.

jyj1289 commented:
저도 이 코드가 맞는지 모르겠습니다...🥲

jyj1289 committed:
docs(#83) : 문서 편집
- enum, user 아스키 문서를 편집했어요.
